core/src/worker/workflow/history_update.rs

```diff
@@ -1877,4 +1891,52 @@ mod tests {
         let seq = next_check_peek(&mut update, 7);
         assert_eq!(seq.last().unwrap().event_id, 11);
     }
+
+    #[tokio::test]
+    async fn wft_fail_on_first_task_with_update() {
+        let mut t = TestHistoryBuilder::default();
+        t.add_by_type(EventType::WorkflowExecutionStarted);
+        t.add_workflow_task_scheduled_and_started();
+        t.add_workflow_task_failed_with_failure(
+            WorkflowTaskFailedCause::Unspecified,
+            Default::default(),
+        );
+        t.add_full_wf_task();
+        let accept_id = t.add_update_accepted("1", "upd");
+        let timer_id = t.add_timer_started("1".to_string());
+        t.add_update_completed(accept_id);
+        t.add_timer_fired(timer_id, "1".to_string());
+        t.add_full_wf_task();
+
+        let mut update = t.as_history_update();
+        let seq = next_check_peek(&mut update, 0);
+        // In this case, we expect to see up to the task with update, since the task failure
+        // should be skipped. This means that the peek of the _next_ task will include the update
+        // and thus properly synthesize the update request with the first activation.
+        assert_eq!(seq.len(), 6);
```

Comment:
Without the fix, this would return 3.

Fun fact - if you have a history without the update (IE: One that just ends after the second full wft), it would have returned 6 already in the past, too, but when you add the update, then it returns early, at 3 (exactly because of the fix). Now this returns 6 in both cases.